rasterio/merge.py

```diff
@@ -403,3 +409,321 @@ def nullcontext(obj):
+    output_count=None,
+    resampling=Resampling.nearest,
+    target_aligned_pixels=False,
+    dst_path=None,
+    dst_kwds=None,
+) -> str:
+    """Merge multiple datasets into a single virtual dataset.
+
+    All files must have the same number of bands, data type, and
+    coordinate reference system.
+
+    Input files are merged in their listed order using the reverse
+    painter's algorithm (default) or another method. If the output file
+    exists, its values will be overwritten by input values.
+
+    Geospatial bounds and resolution of a new output file in the units
+    of the input file coordinate reference system may be provided and
+    are otherwise taken from the first input file.
+
+    Roughly equivalent to GDAL's GDALBuildVrt utility.
+
+    Parameters
+    ----------
+    datasets : list
+        List of dataset objects opened in 'r' mode, filenames or
+        PathLike objects source datasets to be merged.
+    bounds: tuple, optional
+        Bounds of the output image (left, bottom, right, top).  If not
+        set, bounds are determined from bounds of input rasters.
+    res: tuple, optional
+        Output resolution in units of coordinate reference system. If
+        not set, the resolution of the first raster is used. If a single
+        value is passed, output pixels will be square.
+    nodata: float, optional
+        Nodata value to use in output file. If not set, uses the nodata
+        value in the first input raster.
+    background : int or float, optional
+        The background fill value for the VRT.
+    dtype: numpy.dtype or string
+        Data type to use in outputfile. If not set, uses the dtype value
+        in the first input raster.
+    indexes : list of ints or a single int, optional
+        Bands to read and merge
+    output_count: int, optional
+        If using callable it may be useful to have additional bands in
+        the output in addition to the indexes specified for read
+    resampling : Resampling, optional
+        Resampling algorithm used when reading input files.  Default:
+        `Resampling.nearest`.
+    target_aligned_pixels : bool, optional
+        Whether to adjust output image bounds so that pixel coordinates
+        are integer multiples of pixel size, matching the ``-tap``
+        options of GDAL utilities.  Default: False.
+    dst_path : str or PathLike, optional
+        Path of output dataset
+    dst_kwds : dict, optional
+        Dictionary of creation options and other paramters that will be
+        overlaid on the profile of the output dataset.
+
+    Returns
+    -------
+    str
+        XML text describing the virtual dataset (VRT).
+
+    Notes
+    -----
+    The methods of the original merge function could be replaced here
+    by C pixel functions. A TODO?
+
+    """
+    if isinstance(datasets[0], (str, os.PathLike)):
+        dataset_opener = rasterio.open
+    else:
+
+        @contextmanager
+        def nullcontext(obj):
```

Comment:
Since this could be confused with `contextlib.nullcontext`, I propose renaming this to something like:
```suggestion
        def dataset_opener(obj):
```
And removing `dataset_opener = nullcontext` below.

Reply:
Given that rasterio supports python 3.9+, why not use contextlib.nullcontext directly?

Reply:
The function defined here behaves differently from `contextlib.nullcontext`. `nullcontext` returns None whereas this function returns the file handle.

Reply:
contextlib.nullcontext will return whatever you pass as an argument.
https://docs.python.org/3/library/contextlib.html#contextlib.nullcontext

Reply:
@groutr great suggestion! I will make it so.